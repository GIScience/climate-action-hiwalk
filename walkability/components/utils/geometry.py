import geopandas as gpd
import shapely
from pyproj import CRS, Transformer
from shapely.ops import transform

WGS84 = CRS('EPSG:4326')


def get_buffered_aoi(aoi: shapely.MultiPolygon, distance: float) -> shapely.MultiPolygon:
    utm = get_utm_zone(aoi)

    geographic_projection_function = Transformer.from_crs(WGS84, utm, always_xy=True).transform
    wgs84_projection_function = Transformer.from_crs(utm, WGS84, always_xy=True).transform
    projected_aoi = transform(geographic_projection_function, aoi)
    buffered_aoi: shapely.MultiPolygon = projected_aoi.buffer(distance)
    return transform(wgs84_projection_function, buffered_aoi)


def get_utm_zone(aoi: shapely.MultiPolygon) -> CRS:
    return gpd.GeoSeries(data=aoi, crs=WGS84).estimate_utm_crs()
