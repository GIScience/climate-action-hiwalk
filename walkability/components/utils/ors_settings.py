from dataclasses import dataclass

import openrouteservice


@dataclass
class ORSSettings:
    # For future reference maybe check this suggestion: https://gitlab.heigit.org/climate-action/plugins/walkability/-/merge_requests/82#note_61406
    client: openrouteservice.Client

    snapping_rate_limit: int = 100
    snapping_request_size_limit: int = 4999

    directions_rate_limit: int = 40
    directions_waypoint_limit: int = 50

    coordinate_precision: float = 0.000001
    isochrone_max_batch_size: int = 5
