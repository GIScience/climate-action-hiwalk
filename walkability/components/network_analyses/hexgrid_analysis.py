import time
import logging
from climatoology.base.artifact import _Artifact, ContinuousLegendData, create_geojson_artifact
from climatoology.base.computation import ComputationResources
from pathlib import Path

import numpy as np
from pyproj import CRS

from walkability.components.network_analyses.network_analyses import geodataframe_to_graph
from walkability.components.utils.geometry import get_buffered_aoi
from walkability.components.utils.misc import PathCategory


import geopandas as gpd
import h3pandas
import networkx as nx
import pandas as pd
import shapely


from operator import itemgetter
from statistics import fmean

from walkability.components.utils.misc import generate_colors

log = logging.getLogger(__name__)


def hexgrid_permeability_analysis(
    paths: gpd.GeoDataFrame, aoi: shapely.MultiPolygon, max_walking_distance: float, resources: ComputationResources
) -> _Artifact:
    hexgrid_permeability = get_hexgrid_permeability(paths=paths, aoi=aoi, max_walking_distance=max_walking_distance)

    artifact = build_hexgrid_permeability_artifact(hexgrid_permeability=hexgrid_permeability, resources=resources)
    return artifact


def get_hexgrid_permeability(
    paths: gpd.GeoDataFrame, aoi: shapely.MultiPolygon, max_walking_distance: float
) -> gpd.GeoDataFrame:
    begin = time.time()
    log.info('Computing hexgrid permeability')

    buffered_aoi = get_buffered_aoi(aoi, distance=max_walking_distance)
    local_crs = paths.estimate_utm_crs()

    projected_paths = paths.to_crs(local_crs)
    projected_paths = projected_paths[projected_paths['category'] != PathCategory.NOT_WALKABLE]

    destinations = create_destinations(buffered_aoi, local_crs, projected_paths)

    split_paths, snapped_destinations = snap(points=destinations, paths=projected_paths, crs=local_crs)

    graph = geodataframe_to_graph(df=split_paths)

    origins = snapped_destinations[snapped_destinations.to_crs('EPSG:4326').intersects(aoi)]

    detours = []
    index = []
    for cell in origins.iterrows():
        destinations_per_cell = snapped_destinations.clip(cell[1]['buffer'])
        origin = cell[1]['centroids']
        cell_detours = []
        for destination in destinations_per_cell.iterrows():
            try:
                network_distance = nx.dijkstra_path_length(
                    graph,
                    (origin.x, origin.y),
                    (destination[1]['centroids'].x, destination[1]['centroids'].y),
                    weight='length',
                )
                euclidian_distance = origin.distance(destination[1]['centroids'])

                if euclidian_distance == 0:
                    continue
                else:
                    cell_detours.append(network_distance / euclidian_distance)
            except nx.exception.NetworkXNoPath:
                # TODO pass here with no value (watch fmean exceptions if list is empty)
                pass
        if len(cell_detours) == 0:
            detours.append(np.nan)
            index.append(cell[0])
        else:
            detours.append(fmean(cell_detours))
            index.append(cell[0])

    permeability = pd.DataFrame(data={'permeability': detours}, index=index)
    hexgrid_permeability = permeability.h3.h3_to_geo_boundary()

    end = time.time()
    log.info(f'Finished calculating Hexgrid Detour Factor. Took {end-begin}s')
    return hexgrid_permeability


def create_destinations(
    aoi: shapely.MultiPolygon, local_crs: CRS, paths: gpd.GeoDataFrame, resolution: int = 10
) -> gpd.GeoDataFrame:
    """
    Creates a h3 hexgrid for the aoi and drops all cells that do not contain any paths
    ## Params
    :param:`aoi`: `shapely.Multipolygon` to be polyfilled as a hexgrid in `EPSG:4326`
    :param:`local_crs`: `pyproject.CRS` local UTM CRS of the `aoi` and `paths`
    :param:`paths`: `geopandas.GeoDataFrame` of local paths. CRS must match `local_crs`
    ## Returns
    :return:`destinations`: `geopandas.GeoDataFrame` hexgrid of the `aoi` in `local_crs` with added centroids and 200 m buffer
    """
    log.debug(f'Using h3pandas v{h3pandas.version} to get hexgrid for aoi.')  # need to use h3pandas import

    hexgrid = gpd.GeoDataFrame(geometry=[aoi], crs='EPSG:4326').h3.polyfill_resample(resolution)
    local_hexgrid = hexgrid.to_crs(crs=local_crs)

    local_hexgrid['buffer'] = local_hexgrid.buffer(distance=200)
    local_hexgrid['centroids'] = local_hexgrid.centroid
    destinations = local_hexgrid[local_hexgrid.intersects(paths.union_all())]

    return destinations


def snap(points: gpd.GeoDataFrame, paths: gpd.GeoDataFrame, crs: CRS) -> tuple[gpd.GeoDataFrame, gpd.GeoDataFrame]:
    log.debug('Snapping points to path network.')

    points['nearest_point'] = points.apply(find_nearest_point, lines=paths, axis=1)
    points['nearest_point'].set_crs(crs)
    points['snapping_distance'] = points.apply(lambda row: row['nearest_point'].distance(row['centroids']), axis=1)

    snapping_edges = points.apply(lambda row: shapely.LineString([row['centroids'], row['nearest_point']]), axis=1)

    nearest_points = points['nearest_point'].union_all()
    paths['geometry'] = paths.snap(other=nearest_points, tolerance=1)

    paths['geometry'] = paths.apply(lambda row: shapely.ops.split(row['geometry'], splitter=nearest_points), axis=1)
    split_paths = paths.explode().reset_index(drop=True).set_crs(crs)
    for edge in snapping_edges:
        last_row = split_paths.shape[0]
        split_paths.loc[last_row, 'category'] = PathCategory.DESIGNATED
        split_paths.loc[last_row, 'rating'] = 1.0
        split_paths.loc[last_row, 'geometry'] = edge

    return split_paths, points


def find_nearest_point(row: pd.Series, lines: gpd.GeoDataFrame) -> shapely.Point:
    paths_in_cell = lines.clip(row['geometry'])
    point_candidates = []
    for geometry in paths_in_cell['geometry']:
        centroid, point_on_path = shapely.ops.nearest_points(row['centroids'], geometry)
        distance = centroid.distance(point_on_path)
        candidate = {'point': point_on_path, 'distance': distance}
        point_candidates.append(candidate)

    if len(point_candidates) == 0:
        return shapely.Point()

    sorted_candidates = sorted(point_candidates, key=itemgetter('distance'))
    closest_candidate = sorted_candidates[0]
    return closest_candidate['point']


def build_hexgrid_permeability_artifact(
    hexgrid_permeability: gpd.GeoDataFrame, resources: ComputationResources, cmap_name: str = 'YlOrRd'
) -> _Artifact:
    detour = hexgrid_permeability.permeability
    color = generate_colors(color_by=detour, cmap_name=cmap_name)
    legend = ContinuousLegendData(
        cmap_name=cmap_name, ticks={f'{round(detour.min(), ndigits=2)}': 0, f'{round(detour.max(), ndigits=2)}': 1}
    )

    return create_geojson_artifact(
        features=hexgrid_permeability.geometry,
        layer_name='Hexgrid Detours',
        filename='hexgrid_permeability',
        caption=Path('resources/components/network_analyses/hexgrid_permeability/caption.md').read_text(),
        description=Path('resources/components/network_analyses/hexgrid_permeability/description.md').read_text(),
        label=hexgrid_permeability.permeability.round(2).to_list(),
        color=color,
        legend_data=legend,
        resources=resources,
    )
