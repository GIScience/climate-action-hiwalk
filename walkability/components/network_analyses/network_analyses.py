import time
from typing import Callable, Tuple
import geopandas as gpd
import momepy
import numpy as np
from networkx import set_node_attributes
from osmnx import simplify_graph
from pyproj import CRS
import shapely
from shapely import MultiLineString, LineString
from tqdm import tqdm
import networkx as nx
from climatoology.base.artifact import _Artifact
from climatoology.base.computation import ComputationResources

from walkability.components.network_analyses.network_artifacts import build_network_artifacts
from walkability.components.utils.geometry import euclidian_distance, get_utm_zone
import logging


log = logging.getLogger(__name__)


def connectivity_permeability_analyses(
    paths: gpd.GeoDataFrame,
    walkable_distance: float,
    aoi: shapely.MultiPolygon,
    resources: ComputationResources,
    idw_function: Callable[[float], float] = lambda _: 1,
) -> list[_Artifact]:
    log.info('Starting network analyses')
    connectivity_permeability = get_connectivity_permeability(
        paths=paths, walkable_distance=walkable_distance, projected_crs=get_utm_zone(aoi), idw_function=idw_function
    )

    connectivity_permeability = connectivity_permeability.clip(aoi, keep_geom_type=True)

    artifacts = build_network_artifacts(connectivity_permeability=connectivity_permeability, resources=resources)
    log.debug('Completed network analyses')
    return artifacts


def get_connectivity_permeability(
    paths: gpd.GeoDataFrame,
    walkable_distance: float,
    projected_crs: CRS,
    threshold: float = 0.5,
    idw_function: Callable[[float], float] = lambda _: 1,
) -> gpd.GeoDataFrame:
    begin = time.time()
    original_crs = paths.crs
    log.debug(f'Reprojecting geodataframe from {original_crs} to {projected_crs.name}')
    paths = paths.to_crs(projected_crs)

    G = geodataframe_to_graph(paths)
    node_attributes = {}

    def within_max_distance(target: Tuple[float, float]) -> bool:
        return euclidian_distance(center, target) <= walkable_distance

    def within_double_max_distance(target: Tuple[float, float]) -> bool:
        return euclidian_distance(center, target) <= walkable_distance * 2

    def is_walkable(start_node: Tuple[float, float], end_node: Tuple[float, float], target_edge_id: int) -> bool:
        edge_data = G.get_edge_data(start_node, end_node)
        return edge_data.get(target_edge_id).get('rating') > threshold

    for center in tqdm(G.nodes, desc='Computing connectivity for nodes', mininterval=2):
        # Skip node if all edge fails threshold check - premature optimisation, does it even help?
        a = G.edges(nbunch=center, data='rating', default=0)
        if all([k[2] < threshold for k in a]):
            continue

        subgraph = nx.subgraph_view(G, filter_node=within_max_distance, filter_edge=is_walkable)
        subgraph_big = nx.subgraph_view(G, filter_node=within_double_max_distance, filter_edge=is_walkable)
        weighting = {
            target: idw_function(euclidian_distance(center, target)) for target in subgraph.nodes if target != center
        }

        shortest_paths = nx.single_source_dijkstra_path_length(subgraph_big, center, weight='length')
        shortest_paths_connectivity = {
            key: weighting[key] for key, value in shortest_paths.items() if 0 < value < walkable_distance
        }

        straightness = 0.0
        for target in subgraph.nodes:
            if center != target:
                network_dist = shortest_paths.get(target, None)
                euclidean_dist = euclidian_distance(center, target)
                if network_dist is None:
                    straightness += 0.1
                else:
                    straightness += euclidean_dist / network_dist
                if straightness < 0.1:
                    straightness = 0.1

        node_attributes[center] = {
            'connectivity': (sum(shortest_paths_connectivity.values())) / (sum(weighting.values()))
            if len(subgraph.nodes) > 1
            else np.nan,
            'permeability': round(straightness / (len(subgraph.nodes) - 1), 2) if len(subgraph.nodes) > 1 else np.nan,
        }
    nx.set_node_attributes(G, node_attributes)
    log.debug('Finished evaluating connectivity for all nodes')

    edge_attributes = {}
    for s_node, e_node, edge_id in G.edges:
        if is_walkable(s_node, e_node, edge_id):
            connectivity = (G.nodes[s_node]['connectivity'] + G.nodes[e_node]['connectivity']) / 2
            permeability = (G.nodes[s_node]['permeability'] + G.nodes[e_node]['permeability']) / 2
        else:
            connectivity = 0.0
            permeability = 0.0
        edge_attributes[(s_node, e_node, edge_id)] = {'connectivity': connectivity, 'permeability': permeability}

    nx.set_edge_attributes(G, edge_attributes)

    result = momepy.nx_to_gdf(G, points=False)

    end = time.time()
    log.info(f'Finished Connectivity-Permeability Calculation. Took {end-begin}s')

    return result.to_crs(original_crs)[['connectivity', 'permeability', 'geometry']]


def geodataframe_to_graph(df: gpd.GeoDataFrame) -> nx.MultiGraph:
    log.debug('Splitting paths at intersections')
    df_ = split_paths_at_intersections(df)

    log.debug('Creating network graph from paths geodataframe')
    G = momepy.gdf_to_nx(df_, multigraph=True, directed=False, length='length')
    node_data = dict()
    for node in G.nodes():
        node_data[node] = {'x': node[0], 'y': node[1]}
    set_node_attributes(G, node_data)

    log.debug('Simplifying graph by removing intermediate nodes along edges')
    G_ = simplify_graph(G.to_directed(), remove_rings=False, edge_attrs_differ=['rating'])

    log.debug('Finished creating graph')
    return G_.to_undirected()


def split_paths_at_intersections(df):
    df = df.drop(errors='ignore', labels='@other_tags', axis=1)
    df.geometry = df.geometry.apply(lambda geom: MultiLineString([geom]) if isinstance(geom, LineString) else geom)
    df_ = (
        df.assign(
            geometry=df.geometry.apply(
                lambda geom: list(
                    list(map(LineString, zip(geom_part.coords[:-1], geom_part.coords[1:]))) for geom_part in geom.geoms
                )
            )
        )
        .explode('geometry')
        .explode('geometry')
    )
    # PERF: `unary_union` might lead to performance issues ...
    # ... since it creates a single geometry
    # `unary_union`: self-intersection geometries
    # NOTE: All properties of the geodataframe are lost
    # geom: MultiLineString = df.unary_union
    # df_ = gpd.GeoDataFrame(data={'geometry': [geom], 'foo': ["bar"]}, crs=df.crs).explode(index_parts=True)
    df_ = gpd.GeoDataFrame(df_).set_crs(df.crs)
    return df_
