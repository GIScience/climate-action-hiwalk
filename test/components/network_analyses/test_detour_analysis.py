import json
import warnings
from functools import partial
from unittest.mock import patch

import geopandas as gpd
import h3
import h3pandas
import numpy as np
import openrouteservice
import pandas as pd
import pytest
import responses
import responses.matchers
import shapely
from approvaltests import DiffReporter, set_default_reporter, verify
from geopandas.testing import assert_geodataframe_equal
from openrouteservice.exceptions import ApiError
from pandas.testing import assert_frame_equal, assert_series_equal
from requests.exceptions import HTTPError, RetryError
from vcr import use_cassette

from walkability.components.network_analyses.detour_analysis import (
    batch_and_filter_spurs,
    batching,
    check_contains_cell,
    create_destinations,
    generate_waypoint_pairs,
    get_cell_distance,
    get_detour_factors,
    get_i_or_j_spurs,
    get_ij_spurs,
    get_ors_walking_distances,
    match_ors_distance_to_cells,
    ors_request,
    snap_batched_records,
    snap_destinations,
)
from walkability.components.utils.ors_settings import ORSSettings


@pytest.fixture(autouse=True)
def configure_approvaltests():
    set_default_reporter(DiffReporter())


def test_get_detour_factors(
    small_aoi,
    expected_detour_factors,
    snapping_response,
    ors_directions_responses,
):
    assert h3pandas.version is not None

    def request_handling(request, directions: list[dict]) -> tuple[int, dict, str] | Exception:
        payload: dict[str, list] = json.loads(request.body)
        request_length = len(payload['coordinates'])

        status_code: int = 404
        body: dict | None = None
        for index, response in enumerate(directions):
            if len(response['routes'][0]['segments']) == (request_length - 1):
                body = response
                status_code = 200
                break

        if body is None:
            body = {'error': 'not_found'}

        headers = {'request_id': '0'}

        return (status_code, headers, json.dumps(body))

    request_call_back = partial(request_handling, directions=ors_directions_responses['responses'])

    with responses.RequestsMock() as rsps:
        rsps.add(method='POST', url='https://api.openrouteservice.org/v2/snap/foot-walking', json=snapping_response)

        rsps.add_callback(
            method='POST',
            url='https://api.openrouteservice.org/v2/directions/foot-walking/json',
            callback=request_call_back,
        )

        result = get_detour_factors(
            aoi=small_aoi, ors_settings=ORSSettings(openrouteservice.Client(key=''), directions_rate_limit=1000)
        )

        assert_geodataframe_equal(
            result.drop(columns='detour_factor').sort_index(),
            expected_detour_factors.drop(columns='detour_factor').sort_index(),
        )
        # TODO why is this warning in here?
        warnings.warn(f'{result.dtypes}, {expected_detour_factors.dtypes}')
        assert_frame_equal(
            result.drop(columns='geometry').sort_index(),
            expected_detour_factors.drop(columns='geometry').sort_index(),
            check_exact=False,
            atol=0.1,
            rtol=0.1,
        )


def test_create_destinations(default_aoi):
    result: pd.DataFrame = create_destinations(default_aoi)

    hexgrid_size = gpd.GeoDataFrame(geometry=[default_aoi], crs='EPSG:4326').h3.polyfill(10).shape[0]

    assert result.shape[0] > 3 * hexgrid_size

    length_counts = result.groupby(by='spur_id').count()

    length_under_2 = length_counts[length_counts['id'] < 2]
    assert length_under_2.empty

    length_over_50 = length_counts[length_counts['id'] > 50]
    assert length_over_50.empty

    for spur_id in set(result['spur_id']):
        spur: pd.DataFrame = result[result['spur_id'] == spur_id]
        id_sublist = spur['id'].to_list()
        for index, _ in enumerate(id_sublist):
            if index >= len(id_sublist) - 1:
                break
            first_cell = np.array(h3.cell_to_local_ij(origin=id_sublist[0], h=id_sublist[index]))
            second_cell = np.array(h3.cell_to_local_ij(origin=id_sublist[0], h=id_sublist[index + 1]))
            difference = second_cell - first_cell
            assert difference[0] in [0, 1]
            assert difference[1] in [0, 1]
            assert np.any(difference)


@pytest.fixture
def default_hexgrid(default_aoi) -> gpd.GeoDataFrame:
    hexgrid = gpd.GeoDataFrame(geometry=[default_aoi], crs='EPSG:4326').h3.polyfill_resample(10).reset_index()
    origin_id = hexgrid.loc[0, 'h3_polyfill']
    hexgrid['local_ij'] = hexgrid['h3_polyfill'].apply(lambda cell_id: h3.cell_to_local_ij(origin=origin_id, h=cell_id))
    hexgrid['local_i'] = hexgrid['local_ij'].apply(lambda ij: ij[0])
    hexgrid['local_j'] = hexgrid['local_ij'].apply(lambda ij: ij[1])
    return hexgrid


@pytest.fixture
def default_origin_id(default_hexgrid):
    return default_hexgrid.loc[0, 'h3_polyfill']


def test_get_i_or_j_spurs(default_aoi, default_hexgrid, default_origin_id):
    min_ij: tuple[int, int] = (default_hexgrid.local_i.min(), default_hexgrid.local_j.min())
    max_ij = (default_hexgrid.local_i.max(), default_hexgrid.local_j.max())

    result_i: pd.DataFrame = get_i_or_j_spurs(
        aoi=default_aoi,
        origin_id=default_origin_id,
        min_ij=min_ij,
        max_ij=max_ij,
        current_value=min_ij[1],
        current_direction='i',
    )
    result_j: pd.DataFrame = get_i_or_j_spurs(
        aoi=default_aoi,
        origin_id=default_origin_id,
        min_ij=min_ij,
        max_ij=max_ij,
        current_value=min_ij[0],
        current_direction='j',
    )

    for spur_id in result_i['spur_id']:
        spur = result_i[result_i['spur_id'] == spur_id]
        id_sublist = spur['id'].to_list()
        for index, _ in enumerate(id_sublist):
            if index >= len(id_sublist) - 1:
                break

            first_cell = np.array(h3.cell_to_local_ij(origin=id_sublist[0], h=id_sublist[index]))
            second_cell = np.array(h3.cell_to_local_ij(origin=id_sublist[0], h=id_sublist[index + 1]))
            difference = second_cell - first_cell
            assert difference[0] == 1
            assert difference[1] == 0

    for spur_id in result_j['spur_id']:
        spur = result_j[result_j['spur_id'] == spur_id]
        id_sublist = spur['id'].to_list()
        for index, _ in enumerate(id_sublist):
            if index >= len(id_sublist) - 1:
                break

            first_cell = np.array(h3.cell_to_local_ij(origin=id_sublist[0], h=id_sublist[index]))
            second_cell = np.array(h3.cell_to_local_ij(origin=id_sublist[0], h=id_sublist[index + 1]))
            difference = second_cell - first_cell
            assert difference[0] == 0
            assert difference[1] == 1


def test_get_ij_spurs(default_aoi, default_hexgrid, default_origin_id):
    max_ij: tuple[int, int] = (default_hexgrid.local_i.max(), default_hexgrid.local_j.max())
    min_ij: tuple[int, int] = (default_hexgrid.local_i.min(), default_hexgrid.local_j.min())

    result: pd.DataFrame = get_ij_spurs(
        aoi=default_aoi, hexgrid=default_hexgrid, origin_id=default_origin_id, min_ij=min_ij, max_ij=max_ij
    )

    for spur_id in result['spur_id']:
        spur = result[result['spur_id'] == spur_id]
        id_sublist = spur['id'].to_list()
        for index, _ in enumerate(id_sublist):
            if index >= len(id_sublist) - 1:
                break

            first_cell = np.array(h3.cell_to_local_ij(origin=id_sublist[0], h=id_sublist[index]))
            second_cell = np.array(h3.cell_to_local_ij(origin=id_sublist[0], h=id_sublist[index + 1]))
            difference = second_cell - first_cell
            assert difference[0] == 1
            assert difference[1] == 1


def test_ij_spurs_order(small_aoi):
    hexgrid = gpd.GeoDataFrame(geometry=[small_aoi], crs='EPSG:4326').h3.polyfill_resample(10).reset_index()
    origin_id = hexgrid.loc[0, 'h3_polyfill']
    hexgrid['local_ij'] = hexgrid['h3_polyfill'].apply(lambda cell_id: h3.cell_to_local_ij(origin=origin_id, h=cell_id))
    hexgrid['local_i'] = hexgrid['local_ij'].apply(lambda ij: ij[0])
    hexgrid['local_j'] = hexgrid['local_ij'].apply(lambda ij: ij[1])
    result: pd.DataFrame = get_ij_spurs(
        small_aoi,
        hexgrid,
        origin_id,
        max_ij=(hexgrid.local_i.max(), hexgrid.local_j.max()),
        min_ij=(hexgrid.local_i.min(), hexgrid.local_j.min()),
    )

    verify(result.to_json(indent=4))


def test_check_contains_cell(default_aoi):
    h3_cell_in_default_aoi = '8a1f8d2f492ffff'
    h3_cell_not_in_default_aoi = '8a1faad6992ffff'

    assert check_contains_cell(aoi=default_aoi, cell_id=h3_cell_in_default_aoi)
    assert not check_contains_cell(aoi=default_aoi, cell_id=h3_cell_not_in_default_aoi)


def test_batch_spurs():
    spurs = pd.DataFrame(
        data={
            'id': ['a', 'b', 'c', 'a', 'b', 'a', 'b', 'c', 'd', 'e', 'a', 'b', 'c', 'd', 'a'],
            'spur_id': ['a', 'a', 'a', 'b', 'b', 'c', 'c', 'c', 'c', 'c', 'd', 'd', 'd', 'd', 'e'],
            'ordinal': [0, 1, 2, 0, 1, 0, 1, 2, 3, 4, 0, 1, 2, 3, 0],
        }
    )

    result = batch_and_filter_spurs(spurs, max_waypoint_number=3)

    expected_result = pd.DataFrame(
        data={
            'id': ['a', 'b', 'c', 'a', 'b', 'a', 'b', 'c', 'c', 'd', 'e', 'a', 'b', 'c', 'c', 'd'],
            'spur_id': [
                'a',
                'a',
                'a',
                'b',
                'b',
                'c:0',
                'c:0',
                'c:0',
                'c:1',
                'c:1',
                'c:1',
                'd:0',
                'd:0',
                'd:0',
                'd:1',
                'd:1',
            ],
            'ordinal': [0, 1, 2, 0, 1, 0, 1, 2, 2, 3, 4, 0, 1, 2, 2, 3],
        }
    )

    assert_frame_equal(result, expected_result)


def test_get_cell_distance():
    destinations = pd.DataFrame(
        data={
            'id': ['8a1faa99684ffff', '8a1faa99685ffff', '8a1faa9968effff', '8a1faa9968c7fff'],
            'spur_id': ['ij:2:0', 'ij:2:0', 'ij:2:1', 'ij:2:1'],
            'ordinal': [0, 1, 0, 1],
        }
    )
    result = get_cell_distance(destinations)
    assert pytest.approx(result, abs=1.0) == 131.75


@pytest.fixture
def small_ors_snapping_response():
    return [
        {'locations': [None]},
        {'locations': [{'location': [8.773085, 49.376161], 'name': 'Schulstraße', 'snapped_distance': 114.44}]},
    ]


@use_cassette
def test_snap_destinations():
    destinations = pd.DataFrame(
        data={
            'id': ['8a1faad6992ffff', '8a1faad69927fff', '8a1faad69927fff', '8a1faad6992ffff'],
            'spur_id': ['ij:1', 'ij:1', 'ij:2', 'ij:2'],
            'ordinal': [0, 1, 0, 1],
        }
    )

    expected_results = pd.DataFrame(
        data={'snapped_location': [[8.773085, 49.376161], None], 'snapped_distance': [122.49, None]},
        index=['8a1faad69927fff', '8a1faad6992ffff'],
    ).rename_axis('id')

    results = snap_destinations(
        destinations,
        ors_settings=ORSSettings(
            openrouteservice.Client(base_url='http://localhost:8080/ors'), snapping_request_size_limit=1
        ),
    )

    assert_frame_equal(results, expected_results)


def test_snap_batched_records(small_ors_snapping_response):
    locations = [
        gpd.GeoSeries(
            index=['8a1faad6992ffff', '8a1faad69927fff'],
            data=[
                shapely.Point(8.774708093757534, 49.37706987059154),
                shapely.Point(8.772978584588666, 49.377259809601995),
            ],
            crs='EPSG:4326',
        ).rename_axis('id'),
    ]

    expected_result = pd.DataFrame(
        data={'snapped_location': [None, [8.773085, 49.376161]], 'snapped_distance': [None, 114.44]},
        index=['8a1faad6992ffff', '8a1faad69927fff'],
    ).rename_axis('id')

    with responses.RequestsMock() as rsps:
        rsps.add(
            method='POST',
            url='https://api.openrouteservice.org/v2/snap/foot-walking',
            json={'locations': [None, small_ors_snapping_response[1]['locations'][0]]},
        )

        result = snap_batched_records(
            ors_settings=ORSSettings(openrouteservice.Client(key='')),
            batched_locations=locations,
        )

        assert_frame_equal(result, expected_result)


@use_cassette
def test_snapping_request_fail_bad_gateway():
    coordinates = [gpd.GeoSeries([shapely.Point(1.0, 1.0), shapely.Point(1.1, 1.1)], crs='EPSG:4326')]
    ors_settings = ORSSettings(client=openrouteservice.Client(key='test-key'))
    with pytest.raises(RetryError):
        snap_batched_records(ors_settings, batched_locations=coordinates)


@use_cassette
def test_snapping_request_fail_forbidden():
    coordinates = [gpd.GeoSeries([shapely.Point(1.0, 1.0), shapely.Point(1.1, 1.1)], crs='EPSG:4326')]
    ors_settings = ORSSettings(client=openrouteservice.Client(key='test-key'))
    with pytest.raises(HTTPError):
        snap_batched_records(ors_settings, batched_locations=coordinates)


def test_batching():
    origins = pd.Series([i for i in range(0, 6)])
    batch_size = 2

    expected_result = [pd.Series([0, 1]), pd.Series([2, 3]), pd.Series([4, 5])]
    result = batching(origins, batch_size)  # type: ignore

    for index, batch in enumerate(result):
        assert len(batch) == batch_size
        assert_series_equal(batch, expected_result[index], check_index=False)


@use_cassette
def test_get_ors_walking_distances():
    destinations = pd.DataFrame(
        data={
            'id': ['8a1faad6992ffff', '8a1faad69927fff'],
            'spur_id': ['a', 'a'],
            'ordinal': [0, 1],
            'snapped_location': [[8.773, 49.376], [8.773085, 49.376161]],
            'snapped_distance': [114.44, 114.44],
        },
    )
    result = get_ors_walking_distances(
        ors_settings=ORSSettings(openrouteservice.Client(base_url='http://localhost:8080/ors')),
        cell_distance=150,
        destinations_with_snapping=destinations,
    )

    verify(result)


@pytest.fixture
def ors_directions_request_fail():
    with patch('openrouteservice.directions.directions') as mock:
        mock.side_effect = mock_directions_with_ors_error
        yield mock


def mock_directions_with_ors_error(
    client: ORSSettings, coordinates: list[list[float]], profile: str, geometry: bool
) -> None:
    raise ApiError(status=500)


@pytest.fixture
def mock_sleep():
    with patch('time.sleep') as mock:
        mock.return_value = None


def test_ors_request_fail(ors_directions_request_fail, mock_sleep):
    coordinates = [[1.0, 1.0], [1.1, 1.1]]
    ors_settings = ORSSettings(client=openrouteservice.Client(base_url=''))
    with pytest.raises(ApiError):
        ors_request(ors_settings, coordinates)

    ors_directions_request_fail.assert_called()
    assert ors_directions_request_fail.call_count == 5


@use_cassette
def test_ors_request():
    result, start_time = ors_request(
        ors_settings=ORSSettings(client=openrouteservice.Client(key='')),
        coordinates=[[8.773, 49.376], [8.773085, 49.376161]],
    )
    assert isinstance(start_time, float)
    verify(result)


def test_match_ors_distance_to_cells():
    spur = pd.DataFrame(
        data={
            'id': ['0', '1'],
            'spur_id': ['a', 'a'],
            'ordinal': [0, 1],
            'snapped_location': [[8.773, 49.376], [8.773085, 49.376161]],
            'snapped_distance': [114.44, 114.44],
        },
    )
    distances: list[float] = [134.2]

    result = match_ors_distance_to_cells(spur, distances)

    verify(result)


def test_match_ors_distance_to_cells_edge_cases():
    spur = pd.DataFrame(
        data={
            'id': ['0', '1', '2'],
            'spur_id': ['a', 'a', 'a'],
            'ordinal': [0, 1, 2],
            'snapped_location': [None, [8.773085, 49.376161], [8.773085, 49.376161]],
            'snapped_distance': [np.nan, 114.44, 20.0],
        },
    )
    distances: list[float] = [134.2]

    result = match_ors_distance_to_cells(spur, distances)

    assert result.shape == (2, 2)

    verify(result)


def test_generate_waypoint_pairs():
    spur = pd.DataFrame(
        data={
            'id': ['0', '1', '2', '3', '4', '5', '6'],
            'spur_id': ['a', 'a', 'a', 'a', 'a', 'a', 'a'],
            'ordinal': [2, 3, 4, 5, 6, 7, 8],
            'snapped_location': [
                None,
                [8.773085, 49.376161],
                [8.773085, 49.376161],
                None,
                None,
                [8.773085, 49.376161],
                None,
            ],
            'snapped_distance': [np.nan, 114.44, 20.0, np.nan, np.nan, 10, np.nan],
        },
    )
    expected_result = [(3, 4), (4, 7)]
    result = generate_waypoint_pairs(spur)

    assert result == expected_result
