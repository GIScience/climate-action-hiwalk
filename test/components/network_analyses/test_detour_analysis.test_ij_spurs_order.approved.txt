{
    "id":{
        "0":"8a1faa996bb7fff",
        "1":"8a1faa9968effff",
        "2":"8a1faa99684ffff",
        "3":"8a1faa996847fff",
        "4":"8a1faa9968effff",
        "5":"8a1faa9968c7fff"
    },
    "spur_id":{
        "0":"ij:0",
        "1":"ij:0",
        "2":"ij:2",
        "3":"ij:2",
        "4":"ij:5",
        "5":"ij:7"
    },
    "ordinal":{
        "0":0,
        "1":1,
        "2":0,
        "3":1,
        "4":0,
        "5":0
    }
}
